app/main/controllers/clusters.py

```diff
@@ -0,0 +1,225 @@
+            if cluster_parent is None:
+                # items now just belong to the project -
+                cluster.project.items.append(item)
+                item.cluster_id = None
+            else:
+                cluster_parent.child_items.append(item)
+                # I may not need to explicitly do this
+                item.cluster_id = cluster_parent.id
+            item.update()
+
+        if cluster_parent is None:
+            cluster.project.update()
+        else:
+            cluster_parent.update()
+        cluster.delete()
+
+        return jsonify({
+            'success': True,
+            'deleted': cluster_id
+        })
+
+    '''
+    Update cluster name
+    '''
+    @app.route('/clusters/<int:cluster_id>', methods=['PATCH'])
+    @requires_auth('update:clusters')
+    def update_cluster_name(user_id, cluster_id):
+        cluster = get_authorized_cluster(user_id, cluster_id)
+        body = request.get_json()
+        if body is None:
+            abort(400)
+
+        new_name = body.get('name', None)
+        if new_name is None:
+            abort(400)
+
+        cluster.name = new_name
+        cluster.update()
+
+        return jsonify({
+            'success': True,
+            'cluster': cluster.format()
+        })
+
+    '''
+    Creates a new cluster from scratch (two sources come near each other)
+    '''
+    @app.route('/clusters/create_new', methods=['POST'])
+    @requires_auth('create:clusters')
+    def create_cluster_from_scratch(user_id):
+        # Check to ensure user is authorized to make changes
+
+        body = request.get_json()
+        if body is None:
+            abort(400)
+
+        item1_id = body.get('item1_id', None)
+        item2_id = body.get('item2_id', None)
+        if item1_id is None or item2_id is None:
+            abort(400)
+
+        # Get sources and create them if necessary
+        item1 = get_authorized_item(user_id, item1_id)
+        item2 = get_authorized_item(user_id, item2_id)
+        if item1 is None or item2 is None:
+            abort(400)
+
+        # Clusters should start with same parent cluster
+        if item1.cluster != item2.cluster:
+            abort(400)
+
+        x = body.get('x_position', None)
+        y = body.get('y_position', None)
+        name = body.get('name', None)
+        if not name:
+            abort(400)
+
+        cluster = Cluster(name=name, x_position=x, y_position=y)
+        if item1.cluster is not None:
+            cluster.parent_cluster = item1.cluster
+        else:
+            cluster.project = item1.project
+        cluster.insert()
+        cluster.child_items.append(item1)
+        cluster.child_items.append(item2)
+        cluster.update()
+        status_code = 201
+
+        return jsonify({
+            'success': True,
+            'cluster': cluster.format()
+        }), status_code
+
+    @app.route('/clusters/<int:cluster_id>/sources/<int:item_id>',
+               methods=['PATCH'])
```

Comment:
This endpoint doesn't really indicate what it actually does. Since you're adding to a cluster, it should be a POST `/clusters/{cluster_id}/items`
```suggestion
    @app.route('/clusters/<int:cluster_id>/items',
               methods=['POST'])
```